crates/volta-layout/src/v4.rs

```diff
@@ -0,0 +1,125 @@
+            }
+        }
+        "bin": shim_dir {}
+        "log": log_dir {}
+        "tools": tools_dir {
+            "inventory": inventory_dir {
+                "node": node_inventory_dir {}
+                "npm": npm_inventory_dir {}
+                "pnpm": pnpm_inventory_dir {}
+                "yarn": yarn_inventory_dir {}
+            }
+            "image": image_dir {
+                "node": node_image_root_dir {}
+                "npm": npm_image_root_dir {}
+                "pnpm": pnpm_image_root_dir {}
+                "yarn": yarn_image_root_dir {}
+                "packages": package_image_root_dir {}
+            }
+            "shared": shared_lib_root {}
+            "user": default_toolchain_dir {
+                "bins": default_bin_dir {}
+                "packages": default_package_dir {}
+                "platform.json": default_platform_file;
+            }
+        }
+        "tmp": tmp_dir {}
+        "hooks.json": default_hooks_file;
+        "layout.v4": layout_file;
+    }
+}
+
+impl VoltaHome {
+    pub fn node_image_dir(&self, node: &str) -> PathBuf {
+        path_buf!(self.node_image_root_dir.clone(), node)
+    }
+
+    pub fn npm_image_dir(&self, npm: &str) -> PathBuf {
+        path_buf!(self.npm_image_root_dir.clone(), npm)
+    }
+
+    pub fn npm_image_bin_dir(&self, npm: &str) -> PathBuf {
+        path_buf!(self.npm_image_dir(npm), "bin")
+    }
+
+    pub fn pnpm_image_dir(&self, version: &str) -> PathBuf {
+        path_buf!(self.pnpm_image_root_dir.clone(), version)
+    }
+
+    pub fn pnpm_image_bin_dir(&self, version: &str) -> PathBuf {
+        path_buf!(self.pnpm_image_dir(version), "bin")
+    }
+
+    pub fn yarn_image_dir(&self, version: &str) -> PathBuf {
+        path_buf!(self.yarn_image_root_dir.clone(), version)
+    }
+
+    pub fn yarn_image_bin_dir(&self, version: &str) -> PathBuf {
+        path_buf!(self.yarn_image_dir(version), "bin")
+    }
+
+    pub fn package_image_dir(&self, name: &str) -> PathBuf {
+        path_buf!(self.package_image_root_dir.clone(), name)
+    }
+
+    pub fn default_package_config_file(&self, package_name: &str) -> PathBuf {
+        path_buf!(
+            self.default_package_dir.clone(),
+            format!("{}.json", package_name)
+        )
+    }
+
+    pub fn default_tool_bin_config(&self, bin_name: &str) -> PathBuf {
+        path_buf!(self.default_bin_dir.clone(), format!("{}.json", bin_name))
+    }
+
+    pub fn node_npm_version_file(&self, version: &str) -> PathBuf {
+        path_buf!(
+            self.node_inventory_dir.clone(),
+            format!("node-v{}-npm", version)
+        )
+    }
+
+    pub fn shim_file(&self, toolname: &str) -> PathBuf {
+        // On Windows, shims are created as `<name>.cmd` since they
+        // are thin scripts that use `volta run` to execute the command
+        #[cfg(windows)]
+        let toolname = format!("{}{}", toolname, ".cmd");
+
+        path_buf!(self.shim_dir.clone(), toolname)
+    }
```

Comment:
For posterity, this is the only difference between the v3 and v4 layouts, and this is the diff:

```diff
    pub fn shim_file(&self, toolname: &str) -> PathBuf {
-       path_buf!(self.shim_dir.clone(), executable(toolname))
+       // On Windows, shims are created as `<name>.cmd` since they
+       // are thin scripts that use `volta run` to execute the command
+       #[cfg(windows)]
+       let toolname = format!("{}{}", toolname, ".cmd");
+
+       path_buf!(self.shim_dir.clone(), toolname)
    }
```